bugbug/models/duplicate.py

```diff
@@ -0,0 +1,172 @@
+            feature_cleanup.dll(),
+            feature_cleanup.fileref(),
+            feature_cleanup.url(),
+            feature_cleanup.synonyms(),
+            feature_cleanup.crash(),
+        ]
+
+        for cleanup_function in cleanup_functions:
+            text = cleanup_function(text)
+        return text
+
+    def get_labels(self):
+
+        classes = {}
+        duplicate_ids = (
+            []
+        )  # Only store id's of bugs that have duplicate or are duplicate
+        all_ids = []  # Store all remaining ids
+
+        count = 2500
+        for bug_data in bugzilla.get_bugs():
+
+            if bug_data["duplicates"] != []:
+                duplicate_ids.append(bug_data["id"])
+                for duplicate_bug in bug_data["duplicates"]:
+                    count -= 1
+                    duplicate_ids.append(duplicate_bug)
+                    if bug_data["id"] in self.bugs and int(duplicate_bug) in self.bugs:
+                        classes[str(bug_data["id"]) + "|" + str(duplicate_bug)] = 1
+
+                    if not count:
+                        break
+            if not count:
+                break
+            else:
+                all_ids.append(int(bug_data["id"]))
+
+        all_ids = set(all_ids)
+
+        duplicate_ids = set(duplicate_ids)
+
+        # When the bug has no duplicates, we create dup-nondup labels = 0
+        count = 2500
+        for bug_data in bugzilla.get_bugs():
+            if bug_data["duplicates"] == []:
+                for dup_bug_id in duplicate_ids:
+                    count -= 1
+                    if bug_data["id"] in self.bugs and int(dup_bug_id) in self.bugs:
+                        classes[str(bug_data["id"]) + "|" + str(dup_bug_id)] = 0
+                    if not count:
+                        break
+                if not count:
+                    break
+
+        # Non we map non-dup to non-dup bug. However, these are really
+        # large in number, so I'm restricting the number to 2.5K
+
+        count = 2500
+        for bug_data in bugzilla.get_bugs():
+            if int(bug_data["id"]) in all_ids:
+                for non_dup_bug_id in all_ids:
+                    if int(bug_data["id"]) != non_dup_bug_id:
+                        if (
+                            bug_data["id"] in self.bugs
+                            and int(non_dup_bug_id) in self.bugs
+                        ):
+                            classes[str(bug_data["id"]) + "|" + str(non_dup_bug_id)] = 0
+                        count -= 1
+                    if not count:
+                        break
+
+                if not count:
+                    break
+
+        return classes, [0, 1]
+
+    def get_text(self, bug):
+        return bug["summary"] + " ".join(
+            comment["text"] for comment in bug["comments"][0:1]
+        )
+
+    def get_feature_names(self):
+        return self.extraction_pipeline.named_steps["tfidf"].get_feature_names()
+
+    def classify(
```

Comment:
You shouldn't need to overwrite `classify` once the model is properly integrated with the current architecture, you'll just need to change the function in model.py a bit to support tuples of bugs instead of only single bugs.

Reply:
Ah, sure. I was trying to make sure I don't touch model.py. Will do :+1:

Reply:
We will definitely need to make a few changes to model.py and bug_features.py in order to support a model like this, which is quite different than the other models.